src/app/CommandHandler.h

```diff
@@ -175,13 +180,38 @@ class CommandHandler
     template <typename CommandData>
     CHIP_ERROR AddResponseData(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
     {
-        ConcreteCommandPath path = { aRequestCommandPath.mEndpointId, aRequestCommandPath.mClusterId, CommandData::GetCommandId() };
-        ReturnErrorOnFailure(PrepareCommand(path, false));
-        TLV::TLVWriter * writer = GetCommandDataIBTLVWriter();
-        VerifyOrReturnError(writer != nullptr, CHIP_ERROR_INCORRECT_STATE);
-        ReturnErrorOnFailure(DataModel::Encode(*writer, TLV::ContextTag(to_underlying(CommandDataIB::Tag::kData)), aData));
+        // We should not encode anything when we are not in the correct state, the user must be using CommandHandler incorrectly.
+        VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);
 
-        return FinishCommand(/* aEndDataStruct = */ false);
+        CHIP_ERROR err = TryAddResponseData(aRequestCommandPath, aData);
+        if (err != CHIP_NO_ERROR)
+        {
+            // We have verified the state above, so this call must success since we must be one of the state required by
+            // ResetResponse.
+            RollbackResponse();
+        }
+        return err;
+    }
+
+    /**
+     * API for adding a response, i.e. I will try to encode a data response (response command), if it failed to encode the response
+     * data, it will encode Protocols::InteractionModel::Status::Failure status code instead. The template parameter T is generally
+     * expected to be a ClusterName::Commands::CommandName::Type struct, but any object that can be encoded using the
+     * DataModel::Encode machinery and exposes the right command id will work.
```

Comment:
```suggestion
     * API for adding a response.  This will try to encode a data response
     * (response command), and if that fails will encode a a
     * Protocols::InteractionModel::Status::Failure status response instead.
     *
     * The template parameter T is generally expected to be a
     * ClusterName::Commands::CommandName::Type struct, but any object that can
     * be encoded using the DataModel::Encode machinery and exposes the right
     * command id will work.
```